src/object/srv_ec_aggregate.c

```diff
@@ -334,6 +399,155 @@ agg_sgl_fini(d_sg_list_t *sgl)
 	}
 }
 
+/* Determines if an extent overlaps a cell.
+ */
+static bool
+agg_overlap(unsigned int estart, unsigned int elen, unsigned int cell,
+	    unsigned int k, unsigned int len, daos_off_t stripenum)
+{
+	daos_off_t cell_start = k * len * stripenum + len * cell;
+
+	estart += k * len * stripenum;
+	if (cell_start <= estart && estart < cell_start + len)
+		return true;
+	if (estart <= cell_start && cell_start < estart + elen)
+		return true;
+	return false;
+}
+
+static unsigned int
+agg_count_cells(uint8_t *fcbit_map, uint8_t *tbit_map, unsigned int estart,
+		unsigned int elen, unsigned int k, unsigned int len,
+		unsigned int stripenum, unsigned int *full_cell_cnt)
+{
+	unsigned int i, cell_cnt = 0;
+
+	for (i = 0; i < k; i++) {
+		if (i * len >= estart &&  estart + elen >= (i + 1) * len) {
+			setbit(tbit_map, i);
+			if (full_cell_cnt) {
+				D_ASSERT(fcbit_map);
+				setbit(fcbit_map, i);
+				(*full_cell_cnt)++;
+			}
+			cell_cnt++;
+		} else if (agg_overlap(estart, elen, i, k, len, stripenum)) {
+			if (!isset(tbit_map, i)) {
+				setbit(tbit_map, i);
+				cell_cnt++;
+			}
+		}
+	}
+
+	return cell_cnt;
+}
+
+/* Initializes the object handle of for the object represented by the entry.
+ * No way to do this until pool handle uuid and container handle uuid are
+ * initialized and share to other servers at higher(pool/container) layer.
+ */
+static int
+agg_get_obj_handle(struct ec_agg_entry *entry)
+{
+	struct daos_obj_layout	*layout;
+	struct ec_agg_param	*agg_param;
+	unsigned int		 k = entry->ae_oca->u.ec.e_k;
+	int			 i, j, rc = 0;
+	bool			 opened = false;
+
+	agg_param = container_of(entry, struct ec_agg_param, ap_agg_entry);
+	if (daos_handle_is_inval(entry->ae_obj_hdl)) {
+		rc = dsc_obj_open(agg_param->ap_pool_info.api_cont_hdl,
+				  entry->ae_oid.id_pub, DAOS_OO_RW,
+				  &entry->ae_obj_hdl);
+		if (!rc)
```

Comment:
I saw the object is being closed in agg_reset_entry()? are you sure reset is being called necessary in all cases?

Reply:
The current version closes the current handle (if valid) (in the agg_reset_entry function) each time a new object is encountered in the iteration.  I add a call after completion of the iteration to close the last object opened (thanks for catching this).